APEXMALHAR-1818 SQL Support for converting given SQL statement to APEX DAG

chinmaykolhatkar opened the pull request:
Features implemented are:
1. SELECT STATEMENT
2. INSERT STATEMENT
3. INNER JOIN with non-empty equi join condition
4. WHERE clause
5. SCALAR functions implemented in calcite are ready to use
6. Custom scalar functions can be registered.
7. Endpoint can be File OR Kafka OR Streaming Port for both input and output
8. CSV Data Format implemented for both input and output side.
9. Static loading of calcite JDBC driver.
10. Testing on local as well as cluster mode.

chinmaykolhatkar commented on library/src/main/java/com/datatorrent/lib/transform/TransformOperator.java:
```diff
@@ -56,12 +56,12 @@
   @NotNull
   private Map<String, String> expressionMap = new HashMap<>();
-  private List<String> expressionFunctions = new LinkedList<>();
+  protected List<String> expressionFunctions = new LinkedList<>();
```
expressionFunctions variable needs to be used in extended class. Hence the change.

chinmaykolhatkar replied:
Done.

chinmaykolhatkar commented on library/src/main/java/com/datatorrent/lib/transform/TransformOperator.java:
```diff
@@ -56,12 +56,12 @@
   private transient Map<PojoUtils.Setter, Expression> transformationMap = new HashMap<>();
-  private Class<?> inputClass;
-  private Class<?> outputClass;
+  protected Class<?> inputClass;
```
inputClass and outputClass variables needs to be used in extended class. Hence the change.

chinmaykolhatkar replied:
Done.

julianhyde commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,272 @@
+  {
+    if (!driverLoaded) {
+      try {
+        Class.forName("org.apache.calcite.jdbc.Driver");
```
In modern Java, JDBC drivers get loaded automatically. So maybe you could skip this method.

chinmaykolhatkar replied:
This is required keeping in mind how apex cli loads the application libraries.
calcite-core jar will be provided as apex application dependency which is not present in initial CLASSPATH when apex cli is started. When apex application is dynamically launched from apex cli, that is when application dependencies (including calcite-core jar) dynamically gets added to CLASSPATH using ThreadContext. At this time, the calcite Driver does not get loaded as the classes are loaded on demand. As SQLExecEnvironment is entry point for apex application to use calcite, I've used static block on this class to load the calcite driver.

julianhyde replied:
Are you sure that classes are loaded on demand?

Try taking out that block and see what happens. You might be surprised.

I see you get the connection using DriverManager.getConnection. That is the standard way to connect to JDBC, and should be safe. See http://docs.oracle.com/javase/6/docs/api/java/util/ServiceLoader.html, which is used by JDBC.

chinmaykolhatkar replied:
Hmmm.. I knew that DriverManager loads the Driver and in newer Java Class.forName is not required, but it somehow did not work for me before. But now I tried, and it worked... Very strange.. I must have done something wrong before.

So, as suggested, I'll update the code removing loading of Driver explicitly. Still, I'll keep an eye for that if that happens again (basically the error I used to get before was Calcite driver not found).

Thanks Julian. :)

chinmaykolhatkar replied:
Done.

chinmaykolhatkar replied:
Done.

julianhyde commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,272 @@
+ * easier to construct a DAG using this object.
+ *
+ * Eg.
+ * <p>
```
I don't know whether you intend to generate usable javadoc from these comments, but without using &lt;pre&gt; or similar the format will be screwed up. Also there are raw &lt; and &gt; symbols.

chinmaykolhatkar replied:
Done

chinmaykolhatkar replied:
Done.

julianhyde commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/ExpressionCompiler.java:
```diff
@@ -0,0 +1,111 @@
+    final RexProgram program = programBuilder.getProgram();
+
+    final BlockBuilder builder = new BlockBuilder();
+    final JavaTypeFactoryImpl javaTypeFactory = new JavaTypeFactoryImpl(rexBuilder.getTypeFactory().getTypeSystem());
```
Why not re-use the existing type factory? (Type factories canonize common types, so mixing them is not recommended.)

chinmaykolhatkar replied:
Done.

julianhyde commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/ApexSQLTable.java:
```diff
@@ -0,0 +1,118 @@
+  @Override
+  public RelDataType getRowType(RelDataTypeFactory relDataTypeFactory)
+  {
+    return endpoint.getRowType(relDataTypeFactory);
```
It's possible that you should return relDataType here, if it is not null. 

I'd also rename the field relDataType to rowType, because it represents the schema of each row.

chinmaykolhatkar replied:
Done.

julianhyde commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/ApexSQLTableFactory.java:
```diff
@@ -0,0 +1,64 @@
+  @SuppressWarnings("unchecked")
+  @Override
+  public Table create(SchemaPlus schemaPlus, String name, Map<String, Object> operands, RelDataType relDataType)
+  {
```
See my previous remarks about relDataType. If it is passed in to the TableFactory.create SPI, Calcite expects the table to remember it.

chinmaykolhatkar replied:
Done.

julianhyde commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/ApexStreamType.java:
```diff
@@ -0,0 +1,26 @@
+
+import org.apache.apex.malhar.sql.table.Endpoint;
+
+public interface ApexStreamType
```
javadoc would be useful here. I don't understand the purpose of this class.

chinmaykolhatkar replied:
This as such is not required. I wanted to type a calcite table object to specific stream type for expected implementation. But as it turned out, this is not required. Will remove this time.

chinmaykolhatkar replied:
Done. Removed the file.

julianhyde commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/TupleSchemaRegistry.java:
```diff
@@ -0,0 +1,184 @@
+
+  public String generateCommonJar() throws IOException
+  {
+    File file = new File("/tmp/schemaSQL.jar");
```
This is not re-entrant.

chinmaykolhatkar replied:
Should make this with tempFile. Will make the changes.

tweise replied:
should also not assume /tmp - use File.createTempFile

chinmaykolhatkar replied:
Done.

julianhyde commented on sql/src/main/java/org/apache/apex/malhar/sql/table/Endpoint.java:
```diff
@@ -0,0 +1,53 @@
+
+import com.datatorrent.api.DAG;
+
+public interface Endpoint
```
this class seems important. javadoc would be useful.

chinmaykolhatkar replied:
Done.

julianhyde reviewed (commented):
A few minor comments, but overall it looks great.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/operators/LineReader.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * This operator reads data from given file/folder in line by line fashion.
+ */
+public class LineReader extends AbstractFileInputOperator<byte[]>
```
this seems to be similar to org.apache.apex.malhar.lib.fs.LineByLineFileInputOperator in Malhar. Can you replace this with it.

yogidevendra replied:
Alternatively, we have FSRecordReaderModule in malhar. It can read lines from files in parallel. Usage example is available here: https://github.com/DataTorrent/examples/blob/master/tutorials/recordReader/src/main/java/com/example/recordReader/Application.java

chinmaykolhatkar replied:
I suggest we have a simpler functionality first... This module is a little more convoluted than what we need here. If in future we want a parallel read functionality, we can take that up.
@yogidevendra Please log a Jira ticket for improvement.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/operators/LineWriter.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * This operator extends {@link AbstractFileOutputOperator} to write each tuple as a line in given file.
+ */
+public class LineWriter extends AbstractFileOutputOperator<String>
```
This is similar to org.apache.apex.malhar.lib.fs.GenericFileOutputOperator.StringFileOutputOperator. You can use this operator instead.

chinmaykolhatkar replied:
Done.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+      try {
+        connection.close();
+      } catch (Exception e) {
+        e.printStackTrace();
```
log instead of print.

chinmaykolhatkar replied:
Done.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+  /**
+   * Register custom function with this {@link SQLExecEnvironment}
+   *
+   * @param name Name of the scalar SQL function that needs make available to SQL Statement
```
that is made available to SQL statement.

chinmaykolhatkar replied:
Done.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+   * This is the main method takes SQL statement as input and contructs a DAG using contructs registered with this
+   * {@link SQLExecEnvironment}.
+   *
+   * @param sql SQL statement that should be converted to a sub-DAG.
```
sub-DAG -> DAG?

chinmaykolhatkar replied:
Done.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,771 @@
+  }
+
+  @SuppressWarnings("unchecked")
+  private static void addDateFields(ClassNode classNode, String fieldName, String fieldNameForMethods, String type)
```
Add a comment why maintaining multiple representations for date is required.

chinmaykolhatkar replied:
Done.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+      }
+    }
+
+    if (currentNodeRelInfo.getOutPort() == null) {
```
is it possible to output in two different table from sql? in this case this part needs to be at executeSQL after traverse?

yogidevendra replied:
Any harm in moving it to calling method in either case?

chinmaykolhatkar replied:
As such there is no plan for outputting to 2 different tables. In relation algebraic representation of SQL, there might be multiple input points but a single output point. There is a query I've put similar to this which will verify the possibility of that: https://lists.apache.org/thread.html/4989caa200ea49b5bad1a64f059c44f1a13d2e98a43e297cbeaf6a82@%3Cdev.calcite.apache.org%3E

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+            currentNodeRelInfo.getRelName());
+        Class schema;
+        if (inputStream.getOutRelDataType() != null) {
+          schema = getSchemaForRelDataType(streamName, inputStream.getOutRelDataType());
```
is it possible to have same schema multiple times in the DAG? In this case class can be generated only once. Can be taken up later if possible.

chinmaykolhatkar replied:
Its very much possible. I would say that's an optimization which can be take up later.

chinmaykolhatkar replied:
Agreed. Logged improvement for future: https://issues.apache.org/jira/browse/APEXMALHAR-2310

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+    String rightKeyExpression = null;
+    for (Integer leftKey : leftKeys) {
+      String name = OperatorUtils.getValidFieldName(join.getLeft().getRowType().getFieldList().get(leftKey));
+      leftKeyExpression = (leftKeyExpression == null) ? name : leftKeyExpression + " + " + name;
```
Will this work correctly for integer keys? i.e join with two tables with int columns a=10  b=11 and  a=11 b=10 may return same leftKeyExpression key if join is on a and b columns. It should be string contact with some delimiter.

chinmaykolhatkar replied:
@tushargosavi Order of the keys is preserved while comparison string is generated. i.e. leftKeyExpression will become "a + b"... Hence for the 2 cases that you mentioned.. The result of expression will be 1011 & 1110 which are different.
Please correct me if I've missing anything.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/ApexSQLTableFactory.java:
```diff
@@ -0,0 +1,64 @@
+    if (endpointSystemType.equalsIgnoreCase(Endpoint.EndpointType.FILE.name())) {
+      endpoint = new FileEndpoint();
+    } else if (endpointSystemType.equalsIgnoreCase(Endpoint.EndpointType.KAFKA.name())) {
+      endpoint = new KafkaEndpoint();
```
In future need to have ability to register endPoint dynamically in populateDAG. It is already a Factory, we could add a  register method for EndPoint with a name and Endpoint implementation.

chinmaykolhatkar replied:
Hmmm... Good idea.

chinmaykolhatkar replied:
A follow up jira is logged for this. Thomas also suggested something similar.
@tushargosavi Please share your any specific opinion if you have any on the jira.
https://issues.apache.org/jira/browse/APEXMALHAR-2292

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/ApexSQLTableFactory.java:
```diff
@@ -0,0 +1,64 @@
+
+    MessageFormat mf;
+    String messageFormat = (String)operands.get(MessageFormat.MESSAGE_FORMAT);
+    if (messageFormat.equalsIgnoreCase(MessageFormat.MessageFormatType.CSV.name())) {
```
above comment applies here too.

chinmaykolhatkar replied:
Done.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/table/FileEndpoint.java:
```diff
@@ -0,0 +1,114 @@
+  public RelInfo populateInputDAG(DAG dag, JavaTypeFactory typeFactory)
+  {
+    LineReader fileInput = dag.addOperator(OperatorUtils.getUniqueOperatorName("FileInput"), LineReader.class);
+    fileInput.setDirectory((String)operands.get(FILE_INPUT_DIRECTORY));
```
Not null validations for FILE_INPUT_DIRECTORY

chinmaykolhatkar replied:
I would rather leave that to operator validation. Please let me know there is any different thought process in this ask.

yogidevendra replied:
Can we give some generic way to pass on all the properties from the operands in the end point to respective operator properties? 
For example, fileInput might have some other properties which user wants to set. We should avoid new code for each new operator, new property.

chinmaykolhatkar replied:
Created a Jira ticket for this: https://issues.apache.org/jira/browse/APEXMALHAR-2311

@yogidevendra That's a nice suggestion. Can you opinion about this on the Jira? Thanks.

tushargosavi commented on sql/src/main/java/org/apache/apex/malhar/sql/table/PortEndpoint.java:
```diff
@@ -0,0 +1,124 @@
+/**
+ * This is an implementation of {@link Endpoint} which defined how data should be read/written to a Apex streaming port.
+ */
+public class PortEndpoint implements Endpoint
```
Rename it to StreamEndPoint.

chinmaykolhatkar replied:
Done.

tushargosavi reviewed (commented):
There are some duplicate code across PortEndpointTest and KafkaEndpointTest. Other than these comments changes are fine.

yogidevendra commented on sql/pom.xml:
```diff
@@ -0,0 +1,216 @@
+  <packaging>jar</packaging>
+
+  <properties>
+    <checkstyle.console>false</checkstyle.console>
```
Any specific reason for checkstyle.console = false.

chinmaykolhatkar replied:
Was part of pom while development. Missed removing it. Will remove the property.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+import org.apache.apex.malhar.sql.planner.RelNodeVisitor;
+import org.apache.apex.malhar.sql.schema.ApexSQLTable;
+import org.apache.apex.malhar.sql.table.Endpoint;
+import org.apache.calcite.config.Lex;
```
IMO, org.apache.apex imports should be below org.apache.calcite
Import order is generally decided based on how much control you have on the imported class. Imports on the top being least control and the bottom being most control.

You might need to change the import order definition defined in apex-core/misc/ide-settings if required. You may need separate discussion/ JIRA for import order changes in apex-core.

chinmaykolhatkar replied:
I just went ahead with the checkstyle that is mentioned here: https://github.com/apache/apex-core/blob/master/codestyle-config/src/main/resources/apex_checks.xml This is what is used during build also for verification. That said, the travis and jenkins build are passing I believe the import order mentioned in this file is correct.

If we want to change import order in apex-core, that's a seperate discussion and outside the scope of this PR.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+   */
+  public SQLExecEnvironment registerTable(String name, Endpoint endpoint)
+  {
+    registerTable(name, new ApexSQLTable(schema, name, endpoint));
```
What is the expected behavior if registerTable() gets called without call to withModel()

chinmaykolhatkar replied:
registerTable, registerFunction and withModel are 2 different ways to provide information to Calcite schema. So its absolutely valid to call registerTable without calling withModel.

yogidevendra replied:
Would new ApexSQLTable(null, name, endpoint) give correct result?

chinmaykolhatkar replied:
schema will never be null. schema is initialized to a default value at the beginning of the class.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+    }
+  }
+
+  private FrameworkConfig buildFrameWorkConfig()
```
javadoc for this method

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+    FrameworkConfig config = buildFrameWorkConfig();
+    Planner planner = Frameworks.getPlanner(config);
+    try {
+      SqlNode parse = planner.parse(sql);
```
how about renaming to parsedTree?

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+    Planner planner = Frameworks.getPlanner(config);
+    try {
+      SqlNode parse = planner.parse(sql);
+      SqlNode validate = planner.validate(parse);
```
how about renaming to validatedTree?

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+    try {
+      SqlNode parse = planner.parse(sql);
+      SqlNode validate = planner.validate(parse);
+      RelNode tree = planner.rel(validate).rel;
```
rename to relationalTree?

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,771 @@
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
```
IMO, import order should be:

com.google.
com.apache.
com.apache.apex

chinmaykolhatkar replied:
Nope... Its correct as per https://github.com/apache/apex-core/blob/master/codestyle-config/src/main/resources/apex_checks.xml

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,771 @@
+ */
+public class BeanClassGenerator
+{
+  public static final ImmutableMap<String, Character> PRIMITIVE_TYPES;
```
Do we need PRIMITIVE_TYPES from other classes?

chinmaykolhatkar replied:
I think the list contains all primitives. Still, do you have any other primitive types to be considered?

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,771 @@
+    PRIMITIVE_TYPES = ImmutableMap.copyOf(types);
+  }
+
+  public static byte[] createAndWriteBeanClass(String fqcn, List<TupleSchemaRegistry.SQLFieldInfo> fieldList)
```
javadoc

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,771 @@
+    classNode.superName = "java/lang/Object";
+
+    // add default constructor
+    MethodNode constructorNode = new MethodNode(Opcodes.ACC_PUBLIC, "<init>", "()V", null, null);
```
separate this block as separate method.

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,771 @@
+  }
+
+  @SuppressWarnings("unchecked")
+  private static void addIntGetterNSetter(ClassNode classNode, String fieldName, String fieldNameForMethods,
```
addxxxGetterNSetter method has many lines of code duplicated. Can we reuse common code for all types with common method?

chinmaykolhatkar replied:
I suggest we don't.. In all the method there is a minor variations existing in AST of method. Its best to keep a little code duplication here to retain readability and clarity of code.

siyuanh replied:
Except return opcodes I don't see any other differences. And the this method here is used for all byte, boolean, char and int type which is already confusing. I would prefer Yogi's suggestion

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/operators/FilterTransformOperator.java:
```diff
@@ -0,0 +1,68 @@
+public class FilterTransformOperator extends TransformOperator
+{
+  private String condition;
+  private Expression conditionExpression;
```
transient?

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/operators/InnerJoinOperator.java:
```diff
@@ -0,0 +1,50 @@
+  public void setup(Context.OperatorContext context)
+  {
+    this.setExpiryTime(1L);
+    this.setNoOfBuckets(47000);
```
provide explanation for 47000.

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/operators/OperatorUtils.java:
```diff
@@ -0,0 +1,77 @@
+ */
+package org.apache.apex.malhar.sql.operators;
+
+import org.apache.calcite.rel.type.RelDataTypeField;
```
import order

chinmaykolhatkar replied:
Its correct. The build is passing hence.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelInfo.java:
```diff
@@ -0,0 +1,119 @@
+{
+  private List<Operator.InputPort> inputPorts;
+  private Operator operator;
+  private Operator.OutputPort outPort;
```
If we are assuming single output port. Can we document this as a limitation? What are the possible implications? and any thoughts about how this can be handled (in future work)?

chinmaykolhatkar replied:
Its assumption of Relation algebra that at every relational stage there can be multiple inputs but single output. (Correct me if my understanding is wrong).

With keeping this assumption if relational algebra in mind, this is correct and is not a limitation.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+    }
+  }
+
+  public final RelInfo traverse(RelNode n) throws Exception
```
javadoc

yogidevendra replied:
rename n to node or relNode

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+      currentNodeRelInfo = visitJoin((Join)n, inputStreams);
+    } else {
+      currentNodeRelInfo = defaultValue(n, inputStreams);
+    }
```
Move L132-L166 in separate method.

chinmaykolhatkar replied:
Not sure why we should do that, this is post-traverse processing block which appears correct to be in this method.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelInfo.java:
```diff
@@ -0,0 +1,119 @@
+/**
+ * This object communicates stream and connection data between various stages of relational algebra.
+ */
+public class RelInfo
```
Can we rename it to RelNodeInfo or RelOperatorInfo. It should denote that this info is for specific node/stage of RelTree.

chinmaykolhatkar replied:
That's why RelInfo... Rel here stands for "relation"... A relation is single or composite. I think RelInfo looks fine.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+    }
+  }
+
+  public RelInfo visitAggregate(Aggregate aggregate, List<RelInfo> inputStreams) throws Exception
```
How about abstracting defaultValue into visit method on some base class 
and then having specific implementations for visit on delta, project, filter sub-classes (where ever we have non default implementation)?

For supporting some newer type of RelNode one just have to provide implementation for some common interface or extend some base class and override visit implementation.

chinmaykolhatkar replied:
What's the purpose of that? I don't see that base class being used anywhere else. To me this looks appropriate as there are lesser numbers of files and lesser places to make changes on any further additions.

chinmaykolhatkar replied:
The code is simplified now and made modular as well.. Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+    Map<String, String> expMap = new HashMap<>();
+    for (Pair<RelDataTypeField, RelDataTypeField> pair : Pair.zip(filter.getInput().getRowType().getFieldList(),
+        filter.getRowType().getFieldList())) {
+      String leftName = OperatorUtils.getFieldName(pair.left);
```
Could you please document (Code comments/ javadoc) when leftName would be different from rightName?

chinmaykolhatkar replied:
They're different as per design by Calcite. Not sure what should be documented here.

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+
+  public RelInfo visitTableModify(TableModify modify, List<RelInfo> inputStreams) throws Exception
+  {
+    Preconditions.checkArgument(modify.isInsert(), "Only INSERT allowed for table modify");
```
Please document (code comments/ javadocs) other possible types of modify which are not supported.

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+      rightKeyExpression = (rightKeyExpression == null) ? name : rightKeyExpression + " + " + name;
+    }
+
+    String includeFieldStr = "";
```
L416-L426 can be separate utility method.

chinmaykolhatkar replied:
Not sure about this. I think this is at right place.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+    InnerJoinOperator innerJoin = dag.addOperator(OperatorUtils.getUniqueOperatorName(join.getRelTypeName()),
+        InnerJoinOperator.class);
+    innerJoin.setExpiryTime(1L);
+    innerJoin.setNoOfBuckets(47000);
```
Explanation for 47000

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+ *
+ * Above code will evaluate SQL statement and convert the resultant Relational Algebra to a sub-DAG.
+ */
+public class SQLExecEnvironment
```
As a precaution, all new classes should be @ evolving.

chinmaykolhatkar replied:
Good catch.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,457 @@
+
+  private RelInfo addJoinFilter(Join join, RexNode remaining, RelInfo relInfo)
+  {
+    FilterTransformOperator operator = dag.addOperator(OperatorUtils.getUniqueOperatorName(join.getRelTypeName() +
```
Can we reuse common code w.r.t. visitFilter

chinmaykolhatkar replied:
I wanted to but there is a little difference in 2 cases and I wanted to complete out the functionality first. I suggest we take this up later. But good suggestion.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/ApexSQLTable.java:
```diff
@@ -0,0 +1,120 @@
+
+import com.google.common.collect.ImmutableList;
+
+public class ApexSQLTable implements ScannableTable, StreamableTable
```
javadoc

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/table/KafkaEndpoint.java:
```diff
@@ -0,0 +1,134 @@
+  {
+    KafkaSinglePortInputOperator kafkaInput = dag.addOperator(OperatorUtils.getUniqueOperatorName("KafkaInput"),
+        KafkaSinglePortInputOperator.class);
+    kafkaInput.setTopics((String)operands.get(KAFKA_TOPICS));
```
Same comment as above. No custom code per property.

chinmaykolhatkar replied:
This is a good idea. I suggest we take this up as improvement to apex-calcite integration.

chinmaykolhatkar replied:
https://issues.apache.org/jira/browse/APEXMALHAR-2311 as a follow up task should take care of this.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/table/MessageFormat.java:
```diff
@@ -0,0 +1,87 @@
+  enum MessageFormatType
+  {
+    CSV,
+    JSON
```
Do we have support for JSON as well? If not, should we remove JSON for now?

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/test/java/org/apache/apex/malhar/sql/FileEndpointTest.java:
```diff
@@ -0,0 +1,161 @@
+    defaultTZ = TimeZone.getDefault();
+    TimeZone.setDefault(TimeZone.getTimeZone("GMT"));
+
+    File inputFile = new File("/tmp/input/input.csv");
```
Instead of using /tmp; create a directory under target. Refer https://github.com/apache/apex-malhar/blob/127ffe75226abcf91aa5234713ca5f99bc76c46f/library/src/test/java/com/datatorrent/lib/io/fs/FSInputModuleAppTest.java for example.

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,248 @@
+   * @param model String content of model file.
+   * @return Returns this {@link SQLExecEnvironment}
+   */
+  public SQLExecEnvironment withModel(String model)
```
Should we have withModel(File modelFile) as well?

chinmaykolhatkar replied:
I think at this level its best to keep it as String. I don't think it should be the responsibility of this class to read a file.

yogidevendra commented on sql/src/test/java/org/apache/apex/malhar/sql/FileEndpointTest.java:
```diff
@@ -0,0 +1,161 @@
+      Configuration conf = new Configuration(false);
+      lma.prepareDAG(new Application(model), conf);
+      LocalMode.Controller lc = lma.getController();
+      lc.run(10 * 1000); // runs for 30 seconds and quits
```
this is anti-pattern frequently used in tests. We decided to avoid this in newer tests. 
Test should wait for outputFile.exists() and then compare results with actual results. Read discussion about this on https://github.com/apache/apex-malhar/pull/383 and https://github.com/DataTorrent/examples/pull/16

chinmaykolhatkar replied:
This emits to console output operator. Hence nothing to verify here. Will do it for another test.

tweise replied:
Without assertions, what is the purpose of the test?

chinmaykolhatkar replied:
This is select only statement test. I understand the point that without assertion this does not make sense. But this still give developer to manually verify the results.

tweise replied:
Again, tests need to assert expected vs. actual. Nothing manual.

chinmaykolhatkar replied:
Got it.. In test, I changed stdout to capture the results from ConsoleOutputOperator and then reverted to original stdout.
This way I am able to verify the output from consoleoutputoperator too.

This is done.

yogidevendra commented on sql/src/test/java/org/apache/apex/malhar/sql/FileEndpointTest.java:
```diff
@@ -0,0 +1,161 @@
+      Configuration conf = new Configuration(false);
+      lma.prepareDAG(new ApplicationSelectInsertWithAPI(), conf);
+      LocalMode.Controller lc = lma.getController();
+      lc.run(10 * 1000); // runs for 30 seconds and quits
```
same as above

chinmaykolhatkar replied:
Will do this.

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/test/java/org/apache/apex/malhar/sql/FileEndpointTest.java:
```diff
@@ -0,0 +1,161 @@
+    TimeZone.setDefault(TimeZone.getTimeZone("GMT"));
+
+    File inputFile = new File("/tmp/input/input.csv");
+    FileUtils.writeLines(inputFile, ImmutableList.of("15/02/2016 10:15:00 +0000,1,paint1,11",
```
input.csv can be added under test resources instead of writing to /tmp/input

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/test/java/org/apache/apex/malhar/sql/KafkaEndpointTest.java:
```diff
@@ -0,0 +1,364 @@
+
+public class KafkaEndpointTest
+{
+  private String testTopicData0 = "dataTopic0";
```
static final?

chinmaykolhatkar replied:
Final ok.. Why static?

chinmaykolhatkar replied:
Made final. I don't think this needs to be static.

yogidevendra commented on sql/src/main/java/org/apache/apex/malhar/sql/table/CSVMessageFormat.java:
```diff
@@ -0,0 +1,136 @@
+  @Override
+  public RelInfo populateOutputDAG(DAG dag, JavaTypeFactory typeFactory)
+  {
+    CsvFormatter formatter = dag.addOperator(OperatorUtils.getUniqueOperatorName("CSVFormatter"), CsvFormatter.class);
```
You may set lineDelimiter property from CsvFormatter DelimitedSchema if you new line is not desirable in the results.

chinmaykolhatkar replied:
Done.

yogidevendra commented on sql/src/test/java/org/apache/apex/malhar/sql/KafkaEndpointTest.java:
```diff
@@ -0,0 +1,364 @@
+          "15/02/2016 10:19:00 +0000,5,paint5,15", "15/02/2016 10:10:00 +0000,6,abcde6,16"));
+
+      // TODO: Workaround to add \r\n char to test results because of bug in CsvFormatter which adds new line char.
+      String[] expectedLines = new String[]{"15/02/2016 10:18:00 +0000,15/02/2016 00:00:00 +0000,OILPAINT4\r\n",
```
configure lineDelimiter in schema for CsvFormatter to fix this.

chinmaykolhatkar replied:
Lets keep it as it is... Its a good check in itself.  :)

yogidevendra reviewed (changes requested):
I have given some comments which includes:
- improving readability, code reuse etc.
- improving javadocs, comments
- improvements in test comparisons, arbitrary waits 

Overall, this is great piece of work. If you feel some items from the feedback comments which would take longer; you can mention about it explicitly saying it can be handled later.

davidyan74 commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,273 @@
+      } catch (Exception e) {
+        logger.error("Failed to close calcite connection.", e);
+      }
+    }
```
use try-with-resource if possible here

chinmaykolhatkar replied:
Done.

davidyan74 commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,273 @@
+   */
+  public SQLExecEnvironment registerTable(String name, Endpoint endpoint)
+  {
+    Preconditions.checkArgument(name != null, "Table name cannot be null");
```
use checkNotNull instead. Also check other usages of checkArgument.

chinmaykolhatkar replied:
Done.

davidyan74 commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,273 @@
+      RelNodeVisitor visitor = new RelNodeVisitor(dag, typeFactory);
+      visitor.traverse(relationalTree);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
```
use `throw Throwables.propagate(e)` instead

chinmaykolhatkar replied:
I've done this... But would like to know how different this is from throw new RuntimeException(e);

davidyan74 replied:
The difference is that `throw new RuntimeException(e)` can wrap another `RuntimeException`.

chinmaykolhatkar replied:
Done.

davidyan74 commented on sql/src/main/java/org/apache/apex/malhar/sql/SQLExecEnvironment.java:
```diff
@@ -0,0 +1,273 @@
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    } finally {
+      planner.close();
```
use try-with-resource

chinmaykolhatkar replied:
planner is not implementing AutoCloseable interface. Hence cannot do it.

julianhyde replied:
you're welcome to contribute that as a patch; i created https://issues.apache.org/jira/browse/CALCITE-1392

davidyan74 commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,475 @@
+      }
+      return newSchema.beanClass;
+    } else {
+      throw new RuntimeException("Yet to be implemented");
```
Should this be UnsupportedOperationException

chinmaykolhatkar replied:
Done.

davidyan74 commented on sql/src/test/java/org/apache/apex/malhar/sql/EmbeddedKafka.java:
```diff
@@ -0,0 +1,167 @@
+      messages.add(new String(record.value()));
+    }
+
+    consumer.close();
```
Use try-with-resource because close() may not be called if an exception is thrown before that.

chinmaykolhatkar replied:
Done.

davidyan74 commented on sql/src/test/java/org/apache/apex/malhar/sql/EmbeddedKafka.java:
```diff
@@ -0,0 +1,167 @@
+      BROKERPORT = Integer.toString(serverSocket.getLocalPort());
+      serverSocket.close();
+    } catch (IOException e) {
+      e.printStackTrace();
```
exception handling

chinmaykolhatkar replied:
Done.

davidyan74 commented on sql/src/test/java/org/apache/apex/malhar/sql/FileEndpointTest.java:
```diff
@@ -0,0 +1,182 @@
+    boolean fileExists;
+    long now = System.currentTimeMillis();
+    Path outDir = new Path("file://" + new File(outputFolder).getAbsolutePath());
+    FileSystem fs = FileSystem.newInstance(outDir.toUri(), new Configuration());
```
use try with resource

chinmaykolhatkar replied:
Done.

davidyan74 commented on sql/src/test/java/org/apache/apex/malhar/sql/KafkaEndpointTest.java:
```diff
@@ -0,0 +1,364 @@
+
+      lc.shutdown();
+    } catch (Exception e) {
+      e.printStackTrace();
```
printStackTrace should not be used in code to be checked in, even for tests

chinmaykolhatkar replied:
Done.

davidyan74 commented on sql/src/test/resources/model/model_file_csv.json:
```diff
@@ -0,0 +1,27 @@
+{
+  version: '1.0',
```
Should we quote these properly?

chinmaykolhatkar replied:
Nope.. That's calcite specific format for model file. Its not exactly json.. Its more of quasi json one.

julianhyde replied:
Calcite uses Jackson in a lenient mode (e.g. it allows unquoted attribute names and comments). But you are welcome to use strict JSON.

davidyan74 replied:
Since Calcite supports strict json, I think we should change this to use strict json so that github will not flag these.

chinmaykolhatkar replied:
Done. Made it strict json.

davidyan74 reviewed (changes requested):
This is a lot of code you put in. :)
I only put in code style comment. I need more understanding on your implementation to give a more constructive review. Maybe we can do a call with @tweise sometime?

tweise commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,796 @@
+ * Creates a bean class on fly.
+ */
+@InterfaceStability.Evolving
+public class BeanClassGenerator
```
As discussed, please move to separate commit with appropriate attribution. It probably should be in lib or does it require extra dependencies?

chinmaykolhatkar replied:
Done

tweise commented on sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java:
```diff
@@ -0,0 +1,475 @@
+
+    RelInfo currentNodeRelInfo;
+
+    if (relNode instanceof Aggregate) {
```
Could we use a map relNode.getClass -> visitor/translator with a common interface instead of the if-then-else? That would also allow for override of visitors for customization.

julianhyde replied:
RewriteRelVisitor https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java#L701 uses a dispatcher (that finds methods via reflection). You could use that approach.

That said, in Calcite you rarely need a visitor. Calcite's algebraic approach strives to map all operations onto a small number of basic operations. You can add new relational operators but core ones account for the vast majority of cases. If you need something truly extensible and composable you should probably go a step further and write a planner rule(s).

chinmaykolhatkar replied:
This is done now. The code is made modular and simpler to follow.

tweise commented on sql/src/test/java/org/apache/apex/malhar/sql/EmbeddedKafka.java:
```diff
@@ -0,0 +1,169 @@
+import kafka.utils.ZkUtils;
+import kafka.zk.EmbeddedZookeeper;
+
+public class EmbeddedKafka
```
This belongs into malhar-kafka. Let's find out how we get it there (whether it should replace what is there currently). @siyuanh

chinmaykolhatkar replied:
Sent a mail about adding this file to malhar-kafka module to dev@apex about this.
@tweise @siyuanh Please share your opinion.

chinmaykolhatkar replied:
I've moved EmbeddedKafka to malhar-kafka:tests

tweise commented on sql/src/main/java/org/apache/apex/malhar/sql/table/KafkaEndpoint.java:
```diff
@@ -0,0 +1,136 @@
+ * This is an implementation of {@link Endpoint} which defined how data should be read/written from kafka messaging system
+ */
+@InterfaceStability.Evolving
+public class KafkaEndpoint implements Endpoint
```
How about exploring generalization of the endpoint concept as a follow-up? Right now Kafka is hard-wired, but there are many other connectors and they may be custom. Perhaps something like a registry and generic configuration through both, table properties and the Apex configuration file?

chinmaykolhatkar replied:
Yes.. Agreed.. Added jira for it: https://issues.apache.org/jira/browse/APEXMALHAR-2292

tweise reviewed (changes requested):
Great to see this coming along! I don't see big blocking issues so far, will spend some more time on it later this week.

otterc committed:
APEXMALHAR-1818 Adding BeanClassGenerator for dynamically creating class

Commit:
APEXMALHAR-1818 SQL Support for converting given SQL statement to APEX DAG.

Features implemented are:
1. SELECT STATEMENT
2. INSERT STATEMENT
3. INNER JOIN with non-empty equi join condition
4. WHERE clause
5. SCALAR functions implemented in calcite are ready to use
6. Custom scalar functions can be registered.
7. Endpoint can be File OR Kafka OR Streaming Port for both input and output
8. CSV Data Format implemented for both input and output side.
9. Testing on local as well as cluster mode.

chinmaykolhatkar reviewed (commented):
Taken care of review comments except the EmbeddedKafka one. Travis & Jenkins are failing because of https://issues.apache.org/jira/browse/APEXMALHAR-2284

siyuanh commented on kafka/src/test/java/org/apache/apex/malhar/kafka/EmbeddedKafka.java:
```diff
@@ -0,0 +1,169 @@
+import kafka.utils.ZkUtils;
+import kafka.zk.EmbeddedZookeeper;
+
+public class EmbeddedKafka
```
Is it for test only? Can we reuse the code in KafkaOperatorTestBase?

tweise replied:
I thought the idea was to use this instead of the existing launcher in the Kafka tests also?

chinmaykolhatkar replied:
Yes.. I would like to change KafkaOperatorTestBase to use this class. To my understanding, this embedded implementation of kafka is stable of testing purpose.

As thomas rightly said, I'll do that I just thought its outside the scope of this PR, hence I moved the EmbeddedKafka class to malhar-kafka:tests as a part of this PR and as a seperate PR, I'll modify KafkaOperatorTestBase to use this class.

tweise commented on kafka/src/test/java/org/apache/apex/malhar/kafka/EmbeddedKafka.java:
```diff
@@ -0,0 +1,169 @@
+    try {
+      FileUtils.deleteDirectory(new File(KAFKA_PATH));
+    } catch (IOException e) {
+      e.printStackTrace();
```
propagate exception

chinmaykolhatkar replied:
Done.

tweise commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/TupleSchemaRegistry.java:
```diff
@@ -0,0 +1,229 @@
+      try {
+        newSchema.generateBean();
+      } catch (IOException | JSONException e) {
+        throw new RuntimeException("Failed t generate schema", e);
```
typo

chinmaykolhatkar replied:
Done.

tweise commented on sql/src/main/java/org/apache/apex/malhar/sql/schema/TupleSchemaRegistry.java:
```diff
@@ -0,0 +1,229 @@
+      }
+      return newSchema.beanClass;
+    } else {
+      throw new UnsupportedOperationException("Not struct row type is not implemented.");
```
non-struct?

chinmaykolhatkar replied:
Done.

tweise commented on sql/src/test/java/org/apache/apex/malhar/sql/FileEndpointTest.java:
```diff
@@ -0,0 +1,184 @@
+      LocalMode.Controller lc = lma.getController();
+
+      lc.runAsync();
+
```
nit: get rid of extra blank lines (check elsewhere also)

chinmaykolhatkar replied:
testApplication test is to check only select statements... The output of only select statement (i.e. no insert into select) goes to console output operator because there is no insert i.e. destination provided. I want to have assertions there but not sure how to capture results of console output operator.

chinmaykolhatkar replied:
Done.

tweise commented on sql/src/test/java/org/apache/apex/malhar/sql/SerDeTest.java:
```diff
@@ -0,0 +1,223 @@
+        "{\"name\":\"id\",\"type\":\"Integer\"},{\"name\":\"Product\",\"type\":\"String\"}," +
+        "{\"name\":\"units\",\"type\":\"Integer\"}]}";
+
+    Endpoint endpoint = new FileEndpoint("/tmp/test", new CSVMessageFormat(schema));
```
use createTempFile

chinmaykolhatkar replied:
This is the test for dag validation and serde. Its not really having any file or using it. There is no app run here. For clarify I've changed it to "dummyFilePath"

chinmaykolhatkar replied:
Done.

tweise commented on sql/src/test/java/org/apache/apex/malhar/sql/SerDeTest.java:
```diff
@@ -0,0 +1,223 @@
+        "{\"name\":\"Product\",\"type\":\"String\"}]}";
+
+    SQLExecEnvironment.getEnvironment()
+        .registerTable("ORDERS", new FileEndpoint("/tmp/test", new CSVMessageFormat(schemaIn)))
```
see how other tests write to isolated temp files, else this is poison for CI

chinmaykolhatkar replied:
Not a real test. The app does not run here. The test only does dag validation and serde check. Again for clarity, I've changed this to dummyFilePath.

chinmaykolhatkar replied:
Done.

tweise commented on sql/src/test/java/org/apache/apex/malhar/sql/StreamEndpointTest.java:
```diff
@@ -0,0 +1,171 @@
+
+      lc.shutdown();
+    } catch (Exception e) {
+      e.printStackTrace();
```
remove this

chinmaykolhatkar replied:
Done.

siyuanh commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,796 @@
+ * Creates a bean class on fly.
+ */
+@InterfaceStability.Evolving
+public class BeanClassGenerator
```
Can't we just use BeanGenerator from cglib?

julianhyde replied:
@tweise I've not heard it called "fusion" but I like the idea. DAGs should be hybrid. Some steps can be SQL (either a "CREATE VIEW AS SELECT ..." or "INSERT INTO destination SELECT ...") and other steps can be in other languages.

chinmaykolhatkar replied:
@siyuanh I'm not aware of cglib but BeanClassGenerator is tested and working code in other community components like dtGateway. Hence used this. Though I like the idea that you proposed, and we can consider this as an improvement over current one. Let me know if this is fine.

chinmaykolhatkar replied:
@julianhyde @tweise The fusion way is already possible as a part of this PR using StreamEndpoint.
StreamEndpoint can register a apex streaming port as a table. This means.. We can have following cases:
1. Pure style => SQLExecEnvironment
2. Fusion Style => DAG1 -> SQL -> DAG2 (Register output of DAG1 and input of DAG2 as tables)
3. Fusion Style => DAG1 -> SQL (Register output of DAG1 as table and other end i.e. "insert" can be a FileEndpoint/KafkaEndpoint etc...)
4. Fusion Style => SQL -> DAG2 (Register input of DAG2 as table and from input side.. i.e. "select" can be a FileEndpoint/KafkaEndpoint etc...)

The idea of StreamEndpoint gives that flexibility to either have a Pure or fusion style both. 

Test in StreamEndpointTest has examples of that.

tweise replied:
Excellent. And examples will be helpful to users and should land under demos, not hidden in tests.

chinmaykolhatkar replied:
Yes example task is queued: https://issues.apache.org/jira/browse/APEXMALHAR-2304

tweise replied:
I would like some more info on using cglib to generate the classes. If it indeed covers the functionality we need then it is much preferred over pulling in the bean class generator here. Can you please check.

chinmaykolhatkar replied:
Sorry for the long text. but after spending small time I have following information about cglib::

Our requirement:
1. Should be able to create a bean/POJO.
2. Write to jar
3. After jar is present in classpath, should be able to create an object out of it using new or newInstance().

What cglib provides:
The API for cglib are indeed very easy to use.
Looking internally into cglib, it does use ASM5 library to generate the class but also adds some extra callback handlers to it.
From this particular post (http://stackoverflow.com/questions/35146746/write-java-object-into-class-file), we can create a byte[] of generated class to be written to a jar for loading in yarn containers, but along with that it also needs cglib to work. i.e. The generated class is not equivalent of POJO having only private fields+public getters and setters. The bean is not standalone but it contains something more.

Moreover, the bean generated by cglib should be created using BeanGenerator.create method which directly gives a object of the class.

I haven't found any example over the web which uses cglib to generate the bean, write to jar, load jar in different program and use the bean. 
As long all of that is part of the same jvm, cglib seems to be a really good option. But because our functionality is across JVM (apex cli will generate POJOs and those will be used by containers in yarn cluster), my evaluation about this library is not as good option as what we have (BeanClassGenerator).

Please correct me if I'm wrong anywhere.

siyuanh commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,796 @@
+      toStringNode.instructions.add(new FieldInsnNode(Opcodes.GETFIELD, classNode.name, fieldName, fieldJavaType));
+
+      // There is no StringBuilder.append method for short and byte. It takes it as int.
+      if (fieldJavaType.equals("S") || fieldJavaType.equals("B")) {
```
Use Type constant instead

chinmaykolhatkar replied:
Done.

siyuanh commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,796 @@
+   * @throws JSONException
+   */
+  @SuppressWarnings("unchecked")
+  private static void addHashCodeMethod(ClassNode classNode, List<TupleSchemaRegistry.SQLFieldInfo> fieldList)
```
I would prefer bytecode version to be 1.7 so you can use Objects.hash function which is easier for bytecode here. And I believe most hadoop distribution is using 1.7+ by default.

tweise replied:
+1 Apex mandates 1.7 as well.

chinmaykolhatkar replied:
Done.

siyuanh commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,796 @@
+  private static void addEqualsMethod(ClassNode classNode, List<TupleSchemaRegistry.SQLFieldInfo> fieldList)
+    throws JSONException
+  {
+    MethodNode equalsNode = new MethodNode(Opcodes.ACC_PUBLIC, "equals", "(Ljava/lang/Object;)Z", null, null);
```
Same here, Objects.equals is much easier and less error-prone

chinmaykolhatkar replied:
I checked Objects.equals... It has only this: 
    public static boolean equals(Object a, Object b) {
        return (a == b) || (a != null && a.equals(b));
    }

Its does not check for individual elements. Please correct if my understanding is incorrect.

chinmaykolhatkar replied:
@siyuanh Any thought about this? I don't see Objects.equals doing deep equals of elements.
I still suggest we keep existing implementation.

siyuanh commented on sql/src/main/java/org/apache/apex/malhar/sql/codegen/BeanClassGenerator.java:
```diff
@@ -0,0 +1,796 @@
+    return new ByteArrayClassLoader().defineClass(fqcn, inputClass);
+  }
+
+  private static class ByteArrayClassLoader extends ClassLoader
```
ByteArrayClassLoader could be useful (I actually use it in stream module)
Can be pulled up as top level class in lib

chinmaykolhatkar replied:
I didn't exactly created a top level class but rather a ClassLoaderUtils in lib which has this as public static class and also the method which uses this class.

chinmaykolhatkar replied:
Done.

tweise commented on sql/pom.xml:
```diff
@@ -0,0 +1,212 @@
+
+    <!-- For BeanClassGenerator -->
+    <!-- use shaded asm library to avoid conflict -->
+    <dependency>
```
Isn't this an engine dependency? If so, should it be in test or provided scope.

chinmaykolhatkar replied:
asm5 library is required for BeanClassGenerator. Hence included in provided scope.

tweise replied:
I don't see it in provided scope here.

chinmaykolhatkar replied:
Oops... missed that.. Made it in provided scope.
